File: python/DynamicLevenshteinDistanceTechnique.py
from LevenshteinDistanceTechnique import LevenshteinDistanceTechnique

class DynamicLevenshteinDistanceTechnique(LevenshteinDistanceTechnique):

    def name(self):
        return "Dynamic distance"

    def calculateDistance(self, word1, word2):
        lengthWord1 = len(word1)
        lengthWord2 = len(word2)
        table = [[0]*(lengthWord2 + 1) for i in range(lengthWord1 + 1)]
        for i in range(0, lengthWord1 + 1):
            for j in range(0, lengthWord2 + 1):
                if (i == 0):
                    table[i][j] = j
                elif (j == 0):
                    table[i][j] = i
                elif (word1[i - 1] == word2[j - 1]):
                    table[i][j] = table[i-1][j-1]
                else:
                    table[i][j] = 1 + min(min(table[i-1][j-1], table[i][j-1]), table[i-1][j])
        return table[lengthWord1][lengthWord2]

File: ruby/DynamicLevenshteinDistanceTechnique.rb
require "./LevenshteinDistanceTechnique.rb"

class DynamicLevenshteinDistanceTechnique
  include LevenshteinDistanceTechnique

  def name()
    print "Dynamic distance"
  end

  def calculateDistance(word1, word2)
    if (word1.eql?(word2))
      return 0
    end
    table = Array.new(word1.length + 1){Array.new(word2.length + 1)}
    for table.each_with_index do |value, i|
      for table[0].each_with_index do |value, j|
        if (i == 0)
          table[i][j] = j
        elsif (j == 0)
          table[i][j] = i
        elsif (word1[i - 1] == word2[j - 1])
          table[i][j] = table[i - 1][j - 1]
        else
          table[i][j] = 1 + [table[i - 1][j - 1], table[i][j - 1], table[i - 1][j]].min
        end
      end
    end
  end

end
